### PR TITLE
added validators and global exception handlers

### PR DESCRIPTION
**Pull request by NikitaLafinskiy**

### CODE REVIEW

File: src/main/java/com/bookstore/dto/CreateBookRequestDto.java [hunk 1]
```diff
@@ -1,13 +1,30 @@
...
 
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Size;
 import java.math.BigDecimal;
 import lombok.Data;
 
 @Data
 public class CreateBookRequestDto {
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/dto/CreateBookRequestDto.java]:
```suggestion
    @NotBlank
```

File: src/main/java/com/bookstore/dto/CreateBookRequestDto.java [hunk 2]
```diff
@@ -1,13 +1,30 @@
...
+    @Size(min = 5, max = 255)
     private String title;
+
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/dto/CreateBookRequestDto.java]:
```suggestion
    @NotBlank
```

File: src/main/java/com/bookstore/dto/CreateBookRequestDto.java [hunk 3]
```diff
@@ -1,13 +1,30 @@
...
+    @Size(min = 5, max = 255)
     private String author;
+
+    @NotNull
+    @Min(0)
     private BigDecimal price;
+
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/dto/CreateBookRequestDto.java]:
```suggestion
    @NotBlank
```

File: src/main/java/com/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+            HttpHeaders headers,
+            HttpStatusCode status,
+            WebRequest request) {
+        Map<String, Object> error = new HashMap<>();
+        error.put("timestamp", LocalDateTime.now());
+        for (FieldError fieldError : ex.getBindingResult().getFieldErrors()) {
+            error.put(fieldError.getField() + "fieldError",
+                    fieldError.getField() + ": " + fieldError.getDefaultMessage());
+        }
+        return new ResponseEntity<>(error, status);
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/exception/CustomGlobalExceptionHandler.java]: set Bad request status

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
there are changes from previous HW, don't combine several HWs into one PR

File: src/main/java/com/bookstore/dto/CreateBookRequestDto.java [hunk 5]
```diff
@@ -1,13 +1,31 @@
...
+    @NotBlank
+    @Size(min = 5, max = 255)
     private String author;
+
+    @NotNull
+    @Min(0)
     private BigDecimal price;
+
+    @NotBlank
+    private String isbn;
```
Comment (ahoienko) [inline src/main/java/com/bookstore/dto/CreateBookRequestDto.java:27]: Please validate ISBN based on ISBN rules (check ISNB annotation)

File: src/main/java/com/bookstore/entity/Book.java [hunk 6]
```diff
@@ -39,6 +39,6 @@ public class Book {
 
     private String coverImage;
 
-    @Column(nullable = false)
-    private boolean isDeleted;
+    @Column(columnDefinition = "TINYINT(4)", name = "is_deleted")
```
Comment (ahoienko) [inline src/main/java/com/bookstore/entity/Book.java]: What is the reason to make it TINYINT?

File: src/main/java/com/bookstore/exception/CustomGlobalExceptionHandler.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        error.put("timestamp", LocalDateTime.now());
+        for (FieldError fieldError : ex.getBindingResult().getFieldErrors()) {
+            error.put(fieldError.getField() + "fieldError",
+                    fieldError.getField() + ": " + fieldError.getDefaultMessage());
+        }
+        return new ResponseEntity<>(error, HttpStatus.BAD_REQUEST);
+    }
+
+    @ExceptionHandler(Exception.class)
+    public ResponseEntity<Object> handleAll(Exception ex) {
```
Comment (ahoienko) [inline src/main/java/com/bookstore/exception/CustomGlobalExceptionHandler.java]:
```suggestion
    public ResponseEntity<Object> handleException(Exception ex) {
```

File: pom.xml [hunk 8]
```diff
@@ -66,6 +66,11 @@
 			<artifactId>h2</artifactId>
 			<scope>test</scope>
 		</dependency>
+		<dependency>
+			<groupId>org.hibernate.validator</groupId>
+			<artifactId>hibernate-validator</artifactId>
+			<version>8.0.1.Final</version>
+		</dependency>
```
Comment (ahoienko) [inline pom.xml]: Why this dependency is required? Do you use it?

**Review by ahoienko — CHANGES_REQUESTED** [review]
Nice) Lets improve your solution

File: src/main/java/com/bookstore/entity/Book.java (hunk 6 above)
Comment (NikitaLafinskiy) [reply src/main/java/com/bookstore/entity/Book.java]: An error occurs if I get rid of the "columnDefinition" line, I am not sure why but I think it is something related to how hibernate maps fields to the DB columns, probably boolean is represented with a tinyint in mysql, but I changed it to the length of 1 in the new commit

Comment (NikitaLafinskiy) [reply src/main/java/com/bookstore/entity/Book.java]: Schema-validation: wrong column type encountered in column [is_deleted] in table [books]; found [tinyint (Types#TINYINT)], but expecting [bit (Types#BOOLEAN)]

File: src/main/java/com/bookstore/entity/Book.java [hunk 9]
```diff
@@ -39,6 +39,6 @@ public class Book {
 
     private String coverImage;
 
-    @Column(nullable = false)
-    private boolean isDeleted;
+    @Column(columnDefinition = "TINYINT(1)", name = "is_deleted")
+    private Boolean isDeleted = false;
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/Book.java]:
```suggestion
    private boolean isDeleted = false;
```

File: src/main/java/com/bookstore/entity/Book.java [hunk 10]
```diff
@@ -39,6 +39,6 @@ public class Book {
 
     private String coverImage;
 
-    @Column(nullable = false)
-    private boolean isDeleted;
+    @Column(columnDefinition = "TINYINT(1)", name = "is_deleted")
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstore/entity/Book.java]:
```suggestion
    @Column(columnDefinition = "TINYINT(1)")
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**